### PR TITLE
Enhance error handling for notifications from private blogs

### LINKED ISSUE #17127: Some X-Posts fail to load the content when opened from Notifications tab
### Expected behavior

When I see a notification about an X-post in the Notifications tab, and click on it, it lands me on the raw `"X-Post from +site: <title>"` post. If I then click on the `<title>` link on that post, I'm expected to see the actual post content this links to. This works on some posts, but fails on others.

### Actual behavior

For some posts, tapping on the `<title>` link in the raw X-Post post lands to the "Error Loading Post – Open in Browser" page.

At first glance it seems that this only happens with posts having Unicode characters in their title, though @guarani have found some posts without Unicode chars that seem to have the same issue too.

🎬  A video demonstrating the issue is available in p1630920299046800/1630920273.046700-slack-C011BKNU1V5

### Steps to reproduce the behavior

 - Go on the Notifications tab
 - Tap on one notification about an X-Post
 - On the raw X-Post post, tap on the link to the actual post that was cross-linked to the site
 - Check if we can access the linked post's content or if we have an error.
 - Retry with multiple X-Post notifications–especially ones with Unicode or non-standard characters?

##### Tested on [device], iOS [version], WPiOS [version]

iPhone 11 Pro, iOS 14.7.1, WPiOS 18.1.0.3

### PR DESCRIPTION
**Pull request by justtwago**
Fixes #17127 

This PR handles errors for two types of notifications related to private blogs to which the user does not have access: comments and mentions. Mentions are split into mentions either in a comment or a post.

Moreover, the Reader's comments screen displayed an infinite loading progress bar, so it's been fixed too.

**To test:**

First, please temporarily leave a P2 from which you have tons of notifications to make it private for you. I assume it's cdRpT-p2.
Go to Notifications screen and test the following cases
- Open a "mention in/reply to a comment" notification that you've already seen (it has to be cached). You should see the comment on a Mention notification details screen
- Go to the comments screen from the details (tap on the "Comment on >" section on the top of the screen)
- [ ] Make sure you see an error message stating that you cannot access the private blog instead of an infinite loading state.
- [ ] Go back to the notifications screen and open a "mention in/reply to a comment" notification that is not cached yet. Make sure you can see a similar error without seeing the comment.
- Go back to the notifications and find a notification mentioning you **in a post**.
- [ ] Tap on it and ensure you see an error message stating you cannot access the private blog.
- [ ] Smoke test other notifications
- [ ] Subscribe back to the P2 and test the cases again to ensure you have access to all the notifications' details

## Regression Notes
1. Potential unintended areas of impact
Reader comments, Notification details screens

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual tests

3. What automated tests I added (or what prevented me from doing so)
There were no initial tests to the legacy code this PR touches

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

**Comment by hassaanelgarem** [comment]
> User cannot access this private blog

Is this string returned from the endpoint, or is it defined locally?

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift [hunk 1]
```diff
@@ -284,18 +310,18 @@ private extension NotificationCommentDetailViewController {
...
+            failure(nil)
             return
         }
 
         if let blog = blog {
             commentService.loadComment(withID: commentID, for: blog, success: { comment in
                 completion(comment)
             }, failure: { error in
-                completion(nil)
+                failure(error)
```
Comment (alpavanoglu) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift:314]: Does this display BE error directly? If so, that won't be localised and may not make sense for the user.

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift [hunk 2]
```diff
@@ -229,16 +232,39 @@ private extension NotificationCommentDetailViewController {
...
+                    guard let self = self else {
+                        return
+                    }
+                    self.showErrorView(error: error)
                 })
+            }, failure: { [weak self] error in
+                guard let self = self else {
+                    return
+                }
+                self.showErrorView(error: error)
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift:241]: [Nit] In the nested closures, we currently redeclare `[weak self]` after it has already been captured as `weak` by an outer closure. For conciseness and based on [this Stackoverflow question](https://stackoverflow.com/questions/38739129/do-capture-lists-of-inner-closures-need-to-redeclare-self-as-weak-or-unowne), we can safely omit the redundant `weak` capture in the inner closures while still avoiding strong reference cycles.

It's okey to keep the redundant `[weak self]`, but initially, I found the method hard to read. Upon a closer look, I realized that the repeated `[weak self]` captures contributed to its perceived complexity.

**Comment by justtwago** [comment]
Thank you folks! I fixed error messages to display localized strings except for the raw API error message. Also removed redundant weak self checks.

File: WordPress/Classes/ViewRelated/Reader/Detail/ReaderDetailCoordinator.swift [hunk 3]
```diff
@@ -778,3 +790,17 @@ extension ReaderDetailCoordinator {
...
 }
+
+// MARK: - Private Definitions
+
+private extension ReaderDetailCoordinator {
+
+    struct Strings {
+        static let fetchDetailFromPrivateBlogErrorMessage = NSLocalizedString(
+            "readerDetailCoordinator.readerDetail.privateBlogErrorMessage",
+            value: "You have no access to the private blog.",
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Reader/Detail/ReaderDetailCoordinator.swift]: [Nit] How about rewording the error message to "You don't have permission to view this private blog"?

File: WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.m [hunk 4]
```diff
@@ -791,15 +792,26 @@ - (void)refreshNoResultsView
...
     }
+    NSString *image = nil;
+    NSString *subtitle = nil;
+    if (self.fetchCommentsError != nil) {
+        image = @"wp-illustration-reader-empty";
+        NSError *error = self.fetchCommentsError;
+        if (error && [error.domain isEqualToString:WordPressComRestApiErrorDomain] && error.code == WordPressComRestApiErrorCodeAuthorizationRequired) {
+            subtitle = NSLocalizedString(@"You have no access to the private blog.",
+                                          @"Error message that informs reader comments from a private blog cannot be fetched.");
+
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.m:803]: [Suggestion] I think this localized string should use the reverse dns format, right?

**Review by salimbraksa — APPROVED** [review]
I've dropped some minor comments, but PR works as expected. 👍

File: WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.m (hunk 4 above)
Comment (justtwago) [reply WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.m:803]: I'm unsure if I can put a key, value, and description to the Objective-C code. All the `NSLocalizedString` usages I found in our app written in Objective-C look similar to this one.

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/274">24.5</a>. This milestone is due in less than 4 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>
